test/compression_cases_test.go

```diff
@@ -0,0 +1,108 @@
+package test
+
+import (
+	"context"
+	"fmt"
+	"testing"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/credentials/insecure"
+	"google.golang.org/grpc/internal/stubserver"
+	testpb "google.golang.org/grpc/interop/grpc_testing"
+	"google.golang.org/grpc/status"
+)
+
+type mockCompressor struct {
+	grpc.Compressor
+}
+
+func newMockCompressor() grpc.Compressor {
+	return &mockCompressor{grpc.NewGZIPCompressor()}
+}
+
+func (c *mockCompressor) Type() string {
+	return "mock"
+}
+
+type mockDecompressor struct {
+	grpc.Decompressor
+}
+
+func newMockDecompressor() grpc.Decompressor {
+	return &mockDecompressor{grpc.NewGZIPDecompressor()}
+}
+
+func (d *mockDecompressor) Type() string {
+	return "mock"
+}
+
+func TestCompressionCases(t *testing.T) {
+	cases := []struct {
+		desc           string
+		clientUseMock  bool
+		serverUseMock  bool
+		expectedStatus codes.Code
+	}{
+		{
+			desc:           "Client and Server use mock compression",
+			clientUseMock:  true,
+			serverUseMock:  true,
+			expectedStatus: codes.OK,
+		},
+		{
+			desc:           "Only Client use mock compression",
+			clientUseMock:  true,
+			serverUseMock:  false,
+			expectedStatus: codes.Unimplemented,
+		},
+		{
+			desc:           "Only Server use mock compression",
+			clientUseMock:  false,
+			serverUseMock:  true,
+			expectedStatus: codes.Internal,
+		},
+	}
+	for i, tc := range cases {
```

Comment:
Use [t.Run()](https://pkg.go.dev/testing#T.Run) to run subtests instead. You can see the following as an example: https://github.com/grpc/grpc-go/blob/1b1230bb69fba5c6c1653602d576b17025fbf50f/balancer/conn_state_evaluator_test.go#L165-L178

This will make go aware of the different subtests which are part of a test. It also allows you to use defer statements for cleaning up.